jobs/ads-incrementality-dap-collector/ads_incrementality_dap_collector/models.py

```diff
@@ -0,0 +1,227 @@
+    appName: str
+    batchDuration: int
+    branches: List[Branch]
+    bucketConfig: dict
+    channel: str
+    featureIds: list[str]
+    proposedEnrollment: int
+    referenceBranch: Optional[str]
+    slug: str
+    targeting: str
+
+    startDate: date
+    endDate: Optional[date]
+    enrollmentEndDate: Optional[date]
+
+    # Experiment results are removed from DAP after 2 weeks, so our window to collect results for
+    # an experiment is up to 2 weeks after the experiment's latest_collectible_batch_end.
+    # However, we don't need to collect every day of those two weeks (this job runs daily). So this
+    # constant defines how many days after latest_collectible_batch_end to go out and collect and write to BQ.
+    COLLECT_RETRY_DAYS = 7
+
+    @classmethod
+    def from_dict(cls, d) -> "NimbusExperiment":
+        """Load an experiment from dict."""
+        converter = cattrs.BaseConverter()
+        converter.register_structure_hook(
+            date,
+            lambda num, _: datetime.fromisoformat(num.replace("Z", "+00:00"))
+            .astimezone(pytz.utc)
+            .date(),
+        )
+        converter.register_structure_hook(
+            Branch,
+            lambda b, _: Branch(
+                slug=b["slug"], ratio=b["ratio"], features=b["features"]
+            ),
+        )
+        return converter.structure(d, cls)
+
+    def latest_collectible_batch_start(self) -> date:
+        latest_collectible_batch_start = self.startDate
+        # If the experiment's start date is today or in the future, return it
+        if latest_collectible_batch_start >= self.todays_date():
+            return latest_collectible_batch_start
+
+        # While the latest_collectible_batch_start variable is before the batch that includes today...
+        while latest_collectible_batch_start <= self.todays_date():
+            # Increment the latest_collectible_batch_start by the batch interval
+            latest_collectible_batch_start = latest_collectible_batch_start + timedelta(
+                seconds=self.batchDuration
+            )
+        # After the loop, we have the batch start date for the batch that includes today.
+        # We need to return the previous batch, which is now complete and ready for collection.
+        return latest_collectible_batch_start - timedelta(seconds=self.batchDuration) - timedelta(seconds=self.batchDuration)
+
+    def latest_collectible_batch_end(self) -> date:
+        return self.latest_collectible_batch_start() + timedelta(
+            seconds=self.batchDuration, days=-1
+        )
+
+    def next_collect_date(self) -> date:
+        return self.latest_collectible_batch_end() + timedelta(days=1)
+
+    def collect_today(self) -> bool:
+        return (
+            self.latest_collectible_batch_end()
+            < self.todays_date()
+            < (
+                self.latest_collectible_batch_end()
+                + timedelta(days=self.COLLECT_RETRY_DAYS)
+            )
+        )
```

Comment:
Can you add tests for these methods?  They are important in determining if the experiment should be collected.

Reply:
Good call. I should have done this much sooner and it would have made it easier/faster to get the logic right. It was waaay trickier than I realized at first. 🏛️ A classic blunder 🏛️ 

So, there is now a test file for `test_batch_interval_logic` which I think covers all the different cases that are possible here. Would love your eyes/thoughts on this if you can see any code paths left un-exercised.